homeassistant/requirements.py

```diff
@@ -207,3 +90,178 @@ def pip_kwargs(config_dir: str | None) -> dict[str, Any]:
+            await int_or_evt.wait()
+
+            # When we have waited and it's UNDEFINED, it doesn't exist
+            # We don't cache that it doesn't exist, or else people can't fix it
+            # and then restart, because their config will never be valid.
+            if (int_or_evt := cache.get(domain, UNDEFINED)) is UNDEFINED:
+                raise IntegrationNotFound(domain)
+
+        if int_or_evt is not UNDEFINED:
+            return cast(Integration, int_or_evt)
+
+        event = cache[domain] = asyncio.Event()
+
+        try:
+            await self._async_process_integration(integration, done)
+        except Exception:
+            del cache[domain]
+            event.set()
+            raise
+
+        cache[domain] = integration
+        event.set()
+        return integration
+
+    async def _async_process_integration(
+        self, integration: Integration, done: set[str]
+    ) -> None:
+        """Process an integration and requirements."""
+        if integration.requirements:
+            await self.async_process_requirements(
+                integration.domain, integration.requirements
+            )
+
+        deps_to_check = [
+            dep
+            for dep in integration.dependencies + integration.after_dependencies
+            if dep not in done
+        ]
+
+        for check_domain, to_check in DISCOVERY_INTEGRATIONS.items():
+            if (
+                check_domain not in done
+                and check_domain not in deps_to_check
+                and any(check in integration.manifest for check in to_check)
+            ):
+                deps_to_check.append(check_domain)
+
+        if not deps_to_check:
+            return
+
+        results = await asyncio.gather(
+            *(
+                self.async_get_integration_with_requirements(dep, done)
+                for dep in deps_to_check
+            ),
+            return_exceptions=True,
+        )
+        for result in results:
+            if not isinstance(result, BaseException):
+                continue
+            if not isinstance(result, IntegrationNotFound) or not (
+                not integration.is_built_in
+                and result.domain in integration.after_dependencies
+            ):
+                raise result
+
+    async def async_process_requirements(
+        self, name: str, requirements: list[str]
+    ) -> None:
+        """Install the requirements for a component or platform.
+
+        This method is a coroutine. It will raise RequirementsNotFound
+        if an requirement can't be satisfied.
+        """
+        if not (missing := self._find_missing_requirements(requirements)):
+            return
+        self._raise_for_failed_requirements(name, missing)
+
+        async with self.pip_lock:
+            # Recaculate missing again now that we have the lock
+            await self._async_process_requirements(
+                name, self._find_missing_requirements(requirements)
```

Comment:
should we check if missing is empty before passing it to `_async_process_requirements` ?

Reply:
That seems like it would be a rare case but it would be more efficient if it did happen so let's do it

Reply:
https://github.com/home-assistant/core/pull/73451